keps/sig-storage/20191031-storage-capacity-constraints-for-pod-scheduling.md

```diff
@@ -0,0 +1,415 @@
+still end up with a node from time to time where volume creation then
+fails. Rolling back in this case is complicated and outside of the
+scope of this KEP. For example, a pod might use two persistent
+volumes, of which one was created and the other not, and then it
+wouldn’t be obvious whether the existing volume can or should be
+deleted.
+
+For persistent volumes that get created independently of a pod nothing
+changes: it’s still the responsibility of the CSI driver to decide how
+to create the volume and then communicate back through topology
+information where pods using that volume need to run.
+
+Inline volumes could be extended to reference a storage class, which
+then could be used to handle more complex situations (like the LVM
+mirror vs. striped case) also for inline volumes. But this is outside
+the scope of this KEP.
+
+## Proposal
+
+### User Stories
+
+#### Ephemeral PMEM volume for Redis
+
+A [modified Redis server](https://github.com/pmem/redis) can use
+[PMEM](https://pmem.io/) via
+[memkind](http://memkind.github.io/memkind/) as DRAM replacement with
+higher capacity and lower cost at almost the same performance. When it
+starts, all old data is discarded, so an inline ephemeral volume is a
+suitable abstraction for declaring the need for a volume that is
+backed by PMEM and provided by
+[PMEM-CSI](https://github.com/intel/pmem-csi). But PMEM is a resource
+that is local to a node and thus the scheduler has to be aware whether
+enough of it is available on a node before assigning a pod to it.
+
+
+### Caching remaining capacity via the API server
+
+CSI defines the GetCapacity RPC call for the controller service to
+report information about the current capacity but it's not accessible
+in the Kubernetes scheduler. A new API type which represents capacity
+is introduced to solve this (details below).
+
+The CSI driver deployment is responsible for creating and updating
+objects of that type. The external-provisioner gets extended to handle
+this, so CSI drivers only need to implement the GetCapacity call. Each
+capacity value has a lifetime chosen by the creator. Removal of
+expired values is left to the Kubernetes scheduler. This avoids the
+issue that two different processes both try the same operation, which
+would be redundant work. Making the Kubernetes scheduler responsible
+for this has the advantage that it also works when the driver crashes
+or gets de-installed.
+
+The Kubernetes scheduler monitors those objects and uses that
+information when it comes to making scheduling decisions. If the
+scheduler has no current capacity information, it proceeds without it,
+so nothing changes for CSI driver deployments that do not support
+capacity tracking.
+
+This is a problem for deployments that do support it, but haven’t been
+able to provide the necessary information yet. TODO: extend
+CSIDriverInfo with a field that tells the scheduler to retry later
+when information is missing? Avoid the situation by (Somehow? How?!)
```

Comment:
I might have found a solution for this:
- the driver deployment creates the top-level storage info object
- external-provisioner just updates it while the object exists
- removing the driver also removes the object

This way the scheduler knows that if there is an object, it has to schedule based on the information contained in it. The information might still be stale or incomplete, though.

Reply:
Or even simpler: add a CSIDriver.Status field with all the dynamic information.

Reply:
I switched over to that. Conceptually it makes sense to me. But other APIs also would work.